### PR TITLE
feat: kaddht content routing doc

### LINKED ISSUE #23: Write concept doc about Content Routing
Should cover the basics of the content routing interface and how it's used by other parts of libp2p (e.g. rendezvous for AutoRelay).  Should mention that the primary implementation is via kad-dht, but others implementations are possible.

This should be complementary to #13, which covers Peer Routing. Since the content routing interface is distinct from peer routing (even if both are often backed by the same DHT implementation), we should discuss it separately and link between the docs where it makes sense.

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Adds initial Kademlia content routing piece
  > Some of this content may be a bit IPFS-specific, feedback would be great on this too
- resolves #23

## Latest preview

Please view the latest Fleek preview [here](https://bafybeifkqig6oey4ooqtkxbo2wh6o6l3ikwegufnvqut3hxqcnal7y7j3y.on.fleek.co/concepts/content-routing/).

### CODE REVIEW

**Comment by DannyS03** [comment]
Wondering who the best person would be to review this? @p-shahi @mxinden

**Comment by p-shahi** [comment]
@guseggert can you help review this?

File: content/concepts/content-routing.md [hunk 1]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+Peer-to-peer networks are inherently tricky to route because of the following 
+interdisciplinary routing challenges:
 
-Please [refer to this issue](https://github.com/libp2p/docs/issues/23) to track the progress and make suggestions.
+- The lack of universal orchestration that a central server can provide when 
+  querying and retrieving content.
+- Not having a central directory that contains information about reaching every peer 
+  in the network.
+- The presence of high node churn.
+- Data classification is communication-expensive.
```
Comment (guseggert) [inline content/concepts/content-routing.md:14]: I don't know what "data classification" means?

File: content/concepts/content-routing.md [hunk 2]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+- Creating a resilient, scalable, and optimal routing protocol that is resistant to 
+  Sybil attacks and node churn while being future-proof.
```
Comment (guseggert) [inline content/concepts/content-routing.md:16]: A lot of this is restating the previous points, I'd split this into two bullet points:

* Sybil attacks
* Forward compatibility

File: content/concepts/content-routing.md [hunk 3]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+
+Content requests are made in content addressing through the CID.
+If we want to retrieve a file, we send a general request to the network with the CID.
```
Comment (guseggert) [inline content/concepts/content-routing.md:19]:
```suggestion
When retrieving a file using content-addressing, the network identifies the file by its Content ID (CID).
```

File: content/concepts/content-routing.md [hunk 4]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+
+In libp2p, content routing is used for advertising content-addressed chunks of data.
+There exists a content routing interface with operations that allows 
+for content lookups.
```
Comment (guseggert) [inline content/concepts/content-routing.md:23]: I'd remove this if we define "content routing" at the start

File: content/concepts/content-routing.md [hunk 5]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+
+Peer routing schemes may be required to find the peers on the network based on the content 
+routing protocol; DHT-based routing is an example of this, where a DHT is used as a namespace 
+that contains peer information about how to contact different peers on the network. 
+See the peer routing guide for more details.
```
Comment (guseggert) [inline content/concepts/content-routing.md:28]: I'd put this in a note, as it's not really related to Content Routing but is just clearing up some potential confusion, and also link to the guide.

```suggestion
DHTs are also commonly used for resolving Peer IDs to addresses, which is separate from Content Routing. See the Peer Routing Guide for more details.
```

File: content/concepts/content-routing.md [hunk 6]
```diff
@@ -3,6 +3,112 @@ title: Content Routing
...
+interface: KadDHT-libp2p. Libp2p uses an adapation of the 
+[Kademila DHT algorithm](https://pdos.csail.mit.edu/~petar/papers/maymounkov-kademlia-lncs.pdf) 
+to route peers.
+{{% /notice %}}
+
+The DHT provides a key-value store that multiple peers maintain on the network. 
+A different peer stores each row. The network knows which peer is responsible 
+for which row by a distance metric: K-closest neighbor. 
+Learn more about DHT and Kademila on the 
+[IPFS documentation](https://docs.ipfs.tech/concepts/dht/).
```
Comment (guseggert) [inline content/concepts/content-routing.md:45]: I would leave out the discussion of how a DHT works and just link to the DHT docs, like 

```suggestion
For more information about DHTs and the Kademlia implementation, see the [IPFS documentation](https://docs.ipfs.tech/concepts/dht/). 
```

**Review by guseggert — COMMENTED** [review]
I only got halfway through, will followup later with more comments!

**Comment by p-shahi** [comment]
@guseggert any chance you can take a look at this when you have time?

**Comment by DannyS03** [comment]
Closing based on the latest [linked] PRs that incorporate this material as part of the new content routing section and the new docs theme.

**Pull request closed without merging**